compiler/src/dotty/tools/dotc/transform/localopt/RemoveUnnecessaryNullChecks.scala

```diff
@@ -6,104 +6,118 @@ import core.Contexts.Context
-  }
-
-  def isGood(t: Symbol)(implicit ctx: Context): Boolean = {
-    t.exists && initializedVals.contains(t) && {
-      var changed = true
-      var set = Set(t)
-      while (changed) {
-        val oldSet = set
-        set = set ++ set.flatMap(x => checkGood.getOrElse(x, Nil))
-        changed = set != oldSet
-      }
-      !set.exists(x => !initializedVals.contains(x))
-    }
-  }
-
-  def visitor(implicit ctx: Context): Tree => Unit = {
-    case vd: ValDef =>
-      val rhs = vd.rhs
-      if (!vd.symbol.is(Mutable) && !rhs.isEmpty) {
-        def checkNonNull(t: Tree, target: Symbol): Boolean = t match {
-          case Block(_ , expr) =>
-            checkNonNull(expr, target)
-
-          case If(_, thenp, elsep) =>
-            checkNonNull(thenp, target) && checkNonNull(elsep, target)
-
-          case _: New | _: This => true
-
-          case t: Apply if t.symbol.isPrimaryConstructor => true
-
-          case t: Literal => t.const.value != null
-
-          case t: Ident if !t.symbol.owner.isClass =>
-            checkGood.put(target, checkGood.getOrElse(target, Set.empty) + t.symbol)
-            true
-
-          case t: Apply if !t.symbol.owner.isClass =>
-            checkGood.put(target, checkGood.getOrElse(target, Set.empty) + t.symbol)
-            true
-
-          case t: Typed =>
-            checkNonNull(t.expr, target)
-
-          case _ => t.tpe.isNotNull
-
+  val replace: mutable.Map[Tree, Tree] = mutable.Map()
+
+  def clear(): Unit = replace.clear()
+
+  def transformer(implicit ctx: Context): Tree => Tree = 
+    tree => replace.getOrElse(tree, tree)
+
+
+  def visitor(implicit ctx: Context): Tree => Unit =
+    tree => tree match {
+
+      case Block(stats, expr) => 
+        def visitStatements(stats: List[Tree]): Unit =
+          stats match {
+            case s :: tail =>
+              s match {
+                case t: ValDef if !isVar(t.symbol) => 
+                  if (isNotNull(t.rhs)) tail.foreach(flagAsNotNull(t.symbol, _))
+                  else if (isNull(t.rhs)) tail.foreach(flagAsNull(t.symbol, _))
+
+                case Apply(Select(id: Ident, _), _) if !isVar(id.symbol) => 
+                  tail.foreach(flagAsNotNull(id.symbol, _))
+
+                case _ =>
+              }
+              visitStatements(tail)
+            case Nil => 
+          }
+
+        visitStatements(stats :+ expr)
+
+      case If(cond, th, el) => 
+        def visitBranches(cond: Tree, th: Tree, el: Tree): Unit = {
+          cond match {
+            case cond @ Select(expr, _) if cond.symbol == defn.Boolean_! =>
+              visitBranches(expr, el, th) 
+            case cond @ Apply(Select(lhs, _), List(rhs)) if cond.symbol == defn.Boolean_&& =>
+              visitBranches(lhs, th, el)
+              visitBranches(rhs, th, el)
+            case expr => 
+              nullChecksInExpr(expr) match {
+                case Some(symbol) => 
+                  flagAsNull(symbol, th)
```

Comment:
I'm missing something here. I'm I correct saying than `flagAsNull` will flag `Trees` as null/notNull? If so, where do you keep track that this nullness is only valid within this specific branch of the if?